src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java

```diff
@@ -83,22 +86,27 @@ public void addIdentity(String privateKey, final String passphrase, String comme
             }
         }
         try {
-            PEMReader r = new PEMReader(new StringReader(privateKey),
-                    passphrase == null ? null : new PasswordFinder() {
-                        public char[] getPassword() {
-                            return passphrase.toCharArray();
-                        }
-                    });
+            PEMParser r = new PEMParser(new StringReader(privateKey));
+            JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
+            PEMDecryptorProvider decryptionProv = new JcePEMDecryptorProviderBuilder().build(
+                passphrase == null ? null : passphrase.toCharArray());
             try {
                 Object o = r.readObject();
-                if (o instanceof KeyPair) {
-                    agent.getAgent().addIdentity((KeyPair) o, comment);
+                KeyPair keyPair = null;
+
+                if (o instanceof PEMEncryptedKeyPair) {
+                    keyPair = converter.getKeyPair(
+                        ((PEMEncryptedKeyPair) o).decryptKeyPair(decryptionProv));
+                } else if (o instanceof KeyPair) {
+                    keyPair = ((KeyPair) o);
                 }
+                agent.getAgent().addIdentity(keyPair, comment);
             } finally {
                 r.close();
             }
         } catch (Exception e) {
             e.printStackTrace(listener.error(Messages.SSHAgentBuildWrapper_UnableToReadKey(e.getMessage())));
+            e.printStackTrace(listener.getLogger());
```

Comment:
`listener.error` also sends to `getLogger` but with the prefix message and allows for the two to be associated.